### ISSUE TITLE
Update PLY to SLY

### ISSUE THREAD

**Issue by utkonos** [plyara/plyara#15]
dabeaz has rewritten PLY from scratch and modernized everything. The new project is called SLY. It has not hit a stable release yet. However, we should prepare for its stable release and begin our rewrite and refactor soon. More information about SLY on the project page:

https://github.com/dabeaz/sly

**Comment by rshipp** [comment]
> SLY requires the use of Python 3.6 or greater.  Older versions of Python are not supported.

This makes it unusable for me, at least for the time being. I'm not sold on the benefit of migrating just for the sake of it, especially if we lose Python 2 support in the process.

**Comment by utkonos** [comment]
Dropping Python 2 support is a good thing in my opinion. I'm fine with waiting on this change, but there is a wall fast approaching that people need to get ready for. Perhaps we make this change a version 3.0 milestone. Version 1.x and 2.x would use PLY and keep maintenance releases until the death of Python 2. Also, I'm not recommending this for the sake of it. PLY may not be maintained after SLY is ready. I just took a class with dabeaz a few weeks ago and his recommendation is to switch to SLY.

https://pythonclock.org/

I like the quote on the docs site for requests:

"The use of Python 3 is highly preferred over Python 2. Consider upgrading your applications and infrastructure if you find yourself still using Python 2 in production today. If you are using Python 3, congratulations — you are indeed a person of excellent taste." —Kenneth Reitz

**Comment by utkonos** [comment]
SLY is not released as stable yet. I'll keep an eye on it, but I think this change should wait for SLY to be released as stable.

**Comment by utkonos** [comment]
Looks like there's finally a release of SLY. I'll be looking at how to start the migration.

**Comment by malvidin** [comment]
I've written a new core.py and objects.py that uses SLY. 
There are still things to do, but it parses rulesets effectively using the lexing, parsing, and objects based on the official YARA project.
The basic parsing and function names are different enough that calling it plyara isn't accurate.
Would it be more appropriate to create a new slyara project?

**Comment by rshipp** [comment]
Sounds like this is the logical successor. I have no objection to renaming this org/project, so all the old links still work. @utkonos you have owner-permissions here already, and I think you're in the best position to decide this. Let me know if you need anything from me, to update CI/docs or anything.

**Comment by utkonos** [comment]
I have thought about renaming in conjunction with this, but I would like to think long and hard about it.

@malvidin Do you have these changes in a repo somewhere?

The goal I have with this update is to have the current unit tests pass with the new undercarriage. Also, I am going to be attending a training with dabeaz (PLY/SLY author) in about a week and a month. I want to get some insights about SLY from the horse's mouth so-to-speak.

**Comment by malvidin** [comment]
Yes, I just pushed it here, https://github.com/malvidin/plyara/tree/slyara.

There is still things that are not finished.
- tracking imports and externals in the condition
-- test_conditions can't find filename
-- test_detect_dependencies
-- test_detect_imports
- condition_terms is not used
- no JSON export of the rule (test_strings)
- comments are not retained
- raw sections are not retained
- command_line isn't created.

**Comment by utkonos** [comment]
Great! Thanks! I should have some time coming up to wrangle everything together.

**Comment by utkonos** [comment]
@malvidin I would prefer not changing the name of the whole project just because the change in the upstream library. It would cause unnecessary confusion and require people to change their code. Can you make a PR with the SLY code to the SLY branch on plyara? We can get the last parts of the functionality finished up.

Also, with the changes made recently, the code coverage has gone down. One of my goals is to get the unit test coverage back up as high or higher than it was before the recent PR/s.

**Comment by utkonos** [comment]
I have started the steps to build the new 3.0.0 version of plyara based on the SLY library. This is a from the ground up rewrite using a better paradigm of organization of the code. Each step is now separated by functionality. The first part I'm working on is the core data model. The next step will be the tokenizer based on SLY. After that, the export functionality. Three goals with the export functionality:

1. JSON file that matches the export from plyara 2.x
2. New JSON file that may conform to most or all of the 2.x JSON schema components but may deviate to address as many or all of the issues raised at the moment. This format should be informed by the official Go parser and yara-python.
3. Back to a valid YARA rule

I have a new branch for tracking this work here:
https://github.com/plyara/plyara/tree/prep-3.0

The last part will be the parser which may or may not be based on SLY. An alternative is a custom PEG parser.

**Comment by utkonos** [comment]
Minor correction: I am working on the output of valid YARA rules along with the model. This functionality in the 2.x version of plyara can use the most improvement. Therefore I'm solving this problem first.

A good part of the model and output is implemented and correct. Each class and output has a prototype unit test as well.

Comments are more than welcome:
https://github.com/plyara/plyara/tree/prep-3.0

**Comment by utkonos** [comment]
One decision that I've made is to use YARA documentation as the style guide for the output formatting.
